File: src/main/scala/math/cat/Graph.java
package math.cat;

import math.cat.Pair;
import static math.cat.Pair.*;
import static math.cat.Base.*;

import java.util.*;

/**
 * Sample Implementation of (oriented multi-) graph.
 * All code is <a href="http://myjavatools.com/projects/Category/source.zip">here</a>
 */
public abstract class Graph<N, A> extends AbstractSet<N> {
  private Set<N> nodes;

  /*
   * Java technicalities: have to override these methods.
   */
  @Override public Iterator<N> iterator() {  return nodes.iterator(); }
  @Override public int size() { return nodes.size(); }
  @Override public int hashCode() { return nodes.hashCode() * 2 + arrows().hashCode(); }
  @Override public boolean equals(Object o) { return o instanceof Graph && equals((Graph)o); }

  /**
   * Lists all nodes of a graph.
   * @return a set of nodes
   */
  public Set<N> nodes() {
    return Collections.unmodifiableSet(nodes); 
  }

  /**
   * Lists all arrows of a graph.
   * @return a set of arrows
   */
  public abstract Set<A> arrows();

  /**
   * Maps an arrow to its domain, aka source.
   * @param arrow an arrow
   * @return the arrow's domain (source)
   */
  public abstract N d0(A arrow);
  /**
   * Maps an arrow to its codomain, aka target.
   * @param arrow an arrow
   * @return the arrow's codomain (target)
   */
  public abstract N d1(A arrow);

  /**
   * An abstract graph constructor. Takes nodes; arrows are defined in subclasses.
   *
   * @param nodes graph nodes.
   */
  protected Graph(Set<N> nodes) {
    this.nodes = nodes;
    validate();
  }

  /**
   * Validates the graph: checks that for each arrow its domain and codomain is in the graph.
   */
  private void validate() {
    for (A arrow : arrows()) {
      assert nodes.contains(d0(arrow)) : "Domain for " + arrow + " not defined";
      assert nodes.contains(d1(arrow)) : "Codomain for " + arrow + " not defined";
    }
  }

  /**
   * Checks equality of this graph to that one.
   * They are equal if they have the same sets of nodes and arrows, and the arrows
   * originate and end at the same nodes.
   *
   * @param that another graph
   * @return true if they are equal.
   */
  private boolean equals(Graph<N, A> that) {
    boolean isEqual =
        this.nodes().equals(that.nodes()) && // same nodes?
        this.arrows().equals(that.arrows()); // same arrows?

    for (A arrow : arrows()) {
      isEqual = isEqual &&
          this.d0(arrow).equals(that.d0(arrow)) && // same d0?
          this.d1(arrow).equals(that.d1(arrow));   // same d1?
    }
    return isEqual;
  }

  public String toString() {
    StringBuffer out = new StringBuffer();

    for (A arrow : arrows()) {
      if (out.length() > 0) out.append(", ");
      out.append(arrow).append(": ").append(d0(arrow)).append("->").append(d1(arrow));
    }
    return "(" + super.toString() + ", {" + out + "})";
  }

  /**
   * Builds a new graph out of this one, with the inverted arrows.
   * @return a new graph with the the same nodes and with arrows pointing in the opposite directions.
   */
  public Graph<N, A> op() {
    final Graph<N, A> source = this;
    return new Graph<N, A>(nodes) {
      public Set<A> arrows() { return source.arrows(); }
      public N d0(A arrow) { return source.d1(arrow); }
      public N d1(A arrow) { return source.d0(arrow); }
    };
  }

  /**
   * Builds a graph from given nodes and arrows.
   *
   * @param nodes graph nodes
   * @param arrows graph arrows, represented here as mapping arrow tags to (domain,codomain) pairs.
   * @return a new graph
   */
  public static <N, A> Graph<N, A> Graph(Set<N> nodes, final Map<A, Pair<N, N>> arrows) {
    return new Graph<N, A>(nodes) {
      public N d0(A arrow) { return arrows.get(arrow).x(); }
      public N d1(A arrow) { return arrows.get(arrow).y(); }
      public Set<A> arrows() { return arrows.keySet(); }
    };
  }

  /**
   * Builds a graph from given nodes and arrows.
   *
   * @param nodes graph nodes
   * @param d0 maps arrows to their domains
   * @param d1 maps arrows to their codomains
   * @return a new graph
   */
  public static <N, A> Graph<N, A> Graph(Set<N> nodes, final Map<A, N> d0, final Map<A, N> d1) {
    assert d0.keySet().equals(d1.keySet());

    return new Graph<N, A>(nodes) {
      public N d0(A arrow) { return d0.get(arrow); }
      public N d1(A arrow) { return d1.get(arrow); }
      public Set<A> arrows() { return d0.keySet(); }
    };
  }

  /**
   * Builds a graph out of a poset. Arrows are pairs (x,y) where x <= y.
   *
   * @param poset original poset
   * @return graph based on he poset
   */
  public static <N> Graph<N, Pair<N, N>> Graph(PoSet<N> poset) {
    final Set<Pair<N, N>> arrows = new HashSet<Pair<N, N>>();
    for (N x : poset) for (N y : poset) if (poset._le_(x, y)) {
      arrows.add(Pair(x, y));
    }
    return new Graph<N, Pair<N, N>>(poset) {
      public N d0(Pair<N, N> arrow) { return arrow.x(); }
      public N d1(Pair<N, N> arrow) { return arrow.y(); }
      public Set<Pair<N, N>> arrows() { return arrows; }
    };
  }

  public static Graph<String, String> Graph(String string) {
    int splitAt = string.indexOf("], {");
    Map<String, Pair<String, String>> arrows = new HashMap<String, Pair<String, String>>();
    for (String arrow : string.substring(splitAt + 4, string.length() - 2).split(",\\s*")) {
      String[] arrowAndNodes = arrow.split(":\\s*");
      if (arrowAndNodes.length == 2) {
        arrows.put(arrowAndNodes[0], Pair(arrowAndNodes[1].split("\\s*->\\s*")));
      }
    }
    return Graph(parseSet(string.substring(1, splitAt + 1)), arrows);
  }

  public static void main(String[] args) {
    Set<Integer> nodes = Set(0, 1, 2);
    Map<String, Pair<Integer, Integer>> arrows =
        Map(array(    "0.id",    "0.1",     "0.2",     "1.id",      "a",       "b",     "2.1",    "2.id",     "2.a",     "2.b",     "2.swap"),
            array(Pair(0,0), Pair(0,1), Pair(0,2), Pair(1,1), Pair(1,2), Pair(1,2), Pair(2,1), Pair(2,2), Pair(2,2), Pair(2,2), Pair(2,2))
        );
    Graph<Integer,String> three = Graph(nodes, arrows);
    System.out.println(three);
  }
}

File: src/main/scala/math/cat/Graph.scala
package math.cat

import java.io.Reader

import math.cat.Categories.{Cat, read}
import math.sets._
import math.sets.Sets._
import scalakittens.{Good, Result}
import scalakittens.Result._

trait Graph extends GraphData { graph ⇒
  
  def contains(any: Any): Boolean = try { nodes contains any } catch { case _:Exception ⇒ false }

  def size: Int = nodes.size
  
  implicit def pairOfNodes(p: (_, _)): (Node, Node) = (node(p._1), node(p._2))

  override def hashCode: Int = getClass.hashCode + 41 + nodes.hashCode * 61 + arrows.hashCode

  override def equals(x: Any): Boolean = {
    x match {
      case other: Graph ⇒ other.equal(this)
      case _ ⇒ false
    }
  }

  /**
    *  Checks equality of this graph to that one.
    * They are equal if they have the same sets of nodes and arrows, and the arrows
    * originate and end at the same nodes.
    *
    * @param that another graph
    * @return true if they are equal.
    */
  private def equal(that: Graph) = {
    val isEqual = this.nodes == that.nodes && this.arrows == that.arrows
    (isEqual /: arrows) (
      (bool: Boolean, aHere: this.Arrow) ⇒ try {
        val aThere = that.arrow(aHere)
        bool && (d0(aHere) == that.d0(aThere)) && (d1(aHere) == that.d1(aThere))
      } catch { case _: Exception ⇒ false }
    )
  }

  override def toString: String = {
    val nodess = nodes.mkString(", ")
    val arrowss =
      arrows map ((a: Arrow) ⇒ s"$a: ${d0(a)}→${d1(a)}") mkString ", "
    s"({$nodess}, {$arrowss})"
  }

  /**
    * Produces a collection of arrows from x to y.
    *
    * @param from first node
    * @param to   second node
    * @return the set of all arrows from x to y
    */
  def arrowsBetween(from: Node, to: Node): Arrows = asSet(arrows filter ((f: Arrow) ⇒ (d0(f) == from) && (d1(f) == to)))
  
  /**
    * Checks if one arrow follows another
    * @param f an arrow
    * @param g an arrow
    * @return true iff f follows g
    */
  def follows(f: Arrow, g: Arrow): Boolean = {
    d0(f) == d1(g)
  }

  /**
    * Checks if two arrows have the same domain
    * @param f an arrow
    * @param g an arrow
    * @return true iff g and f have the same domain
    */
  def sameDomain(f: Arrow, g: Arrow): Boolean = {
    d0(arrow(f)) == d0(arrow(g))
  }

  /**
    * Checks if two arrows have the same codomain
    * @param f an arrow
    * @param g an arrow
    * @return true iff g and f have the same codomain
    */
  def sameCodomain(f: Arrow, g: Arrow): Boolean = {
    d1(arrow(f)) == d1(arrow(g))
  }

  /**
    * Checks if two arrows are parallel
    * @param f an arrow
    * @param g an arrow
    * @return true iff g and f have the same domain and codomain
    */
  def areParallel(f: Arrow, g: Arrow): Boolean = sameDomain(f, g) && sameCodomain(f, g)

  def unary_~ : Graph = new Graph {
    type Node = graph.Node
    type Arrow = graph.Arrow
    def nodes: Nodes = graph.nodes
    def arrows: Arrows = graph.arrows
    def d0(f: Arrow): Node = graph.d1(f)
    def d1(f: Arrow): Node = graph.d0(f)
  }
  
  def subgraph(aName: String, setOfNodes: Nodes): Graph = {
    require(setOfNodes.subsetOf(nodes), s"Unknown nodes: ${setOfNodes.diff(nodes)}")
    new Graph {
      override val name = aName
      type Node = graph.Node
      type Arrow = graph.Arrow

      def nodes: Nodes = setOfNodes

      def arrows: Arrows = graph.arrows filter (a => setOfNodes(d0(a)) && setOfNodes(d1(a)))

      def d0(f: Arrow): Node = graph.d0(f)

      def d1(f: Arrow): Node = graph.d1(f)
    }
  }
 
  def connectedComponents: Set[Graph] = {
    val connected: BinaryRelation[Node, Node] =
      BinaryRelation((x, y) => arrows.exists(a =>
        (x == d0(a) && y == d1(a)) || (x == d1(a) && y == d0(a))))

    val sets = new FactorSet(nodes, connected)

    sets.zipWithIndex map {
      case (s, i) => subgraph(s"$name.${i+1}", s)
    }
  }

}

private[cat] trait GraphData {
  val name: String = "a graph"
  type Node
  type Arrow
  type Nodes = Set[Node]
  type Arrows = Set[Arrow]
  
  def nodes: Nodes
  def arrows: Arrows
  def d0(f: Arrow): Node
  def d1(f: Arrow): Node

  implicit def node(x: Any): Node = x match {
    case _ if nodes contains x.asInstanceOf[Node] ⇒ x.asInstanceOf[Node]
    case other ⇒
      throw new IllegalArgumentException(s"<<$other>> is not a node in $name")
  }

  implicit def arrow(a: Any): Arrow = a match {
    case _ if arrows contains a.asInstanceOf[Arrow] ⇒ a.asInstanceOf[Arrow]
    case other ⇒
      throw new IllegalArgumentException(s"<<$other>> is not an arrow of $name")
  }

  protected lazy val finiteNodes: Boolean = Sets.isFinite(nodes)
  protected lazy val finiteArrows: Boolean = Sets.isFinite(arrows)
  
  def isFinite: Boolean = finiteNodes && finiteArrows

  def validate: Result[GraphData] =
    OKif(!finiteArrows) orElse {
      Result.fold(arrows map {
        a ⇒
          OKif(nodes contains d0(a), " d0 for " + a + " should be in set of nodes") andAlso
          OKif(nodes contains d1(a), " d1 for " + a + " should be in set of nodes") returning ()
      })
    } returning this
}

object Graph {
  def build[N, A](
    nodes0: Set[N],
    arrows0: Set[A],
    d00: A ⇒ N,
    d10: A ⇒ N): Result[Graph] = {
    val data: GraphData = new GraphData {
      override type Node = N
      override type Arrow = A
      def nodes: Nodes = nodes0
      def arrows: Arrows = arrows0
      def d0(f: Arrow): Node = d00(f)
      def d1(f: Arrow): Node = d10(f)
    }

    data.validate returning
      new Graph {
        def nodes: Nodes = data.nodes.asInstanceOf[Nodes] // TODO: get rid of cast
        def arrows: Arrows = data.arrows.asInstanceOf[Arrows] // TODO: get rid of cast
        override type Node = N
        override type Arrow = A
        override def d0(f: Arrow): Node = d00(f).asInstanceOf[Node]
        override def d1(f: Arrow): Node = d10(f).asInstanceOf[Node]
      }
  }

  def fromArrowMap[N, A] (nodes: Set[N], arrows: Map[A, (N, N)]): Result[Graph] = {
    build(nodes, arrows.keySet, (a:A) ⇒ arrows(a)._1,  (a: A) ⇒ arrows(a)._2)
  }

  def discrete[N](points: Set[N]): Graph =
    new Graph {
      type Node = N
      type Arrow = N
      def nodes: Nodes = points
      def arrows: Arrows = Set.empty
      def d0(f: Arrow): Node = Map.empty(f) // there's nothing there, but we need a signature
      def d1(f: Arrow): Node = Map.empty(f)
    }

  def ofPoset[N](poset: PoSet[N]): Graph = {
    val points = poset.underlyingSet
    val posetSquare = Sets.product2(points, points)
    val goodPairs: Set[(N,N)] = Sets.filter(posetSquare, poset.le)

    new Graph {
      type Node = N
      type Arrow = (N, N)
      def nodes: Nodes = points
      def arrows: Arrows = goodPairs
      def d0(f: Arrow): Node = f._1
      def d1(f: Arrow): Node = f._2
    }
  }        

  class Parser extends Sets.Parser {
    def all: Parser[Result[Graph]] = "("~graph~")" ^^ {case "("~g~")" ⇒ g}

    def graph: Parser[Result[Graph]] = parserOfSet~","~arrows ^^ {case s~","~a ⇒ Graph.fromArrowMap(s, a)}

    def arrows: Parser[Map[String, (String, String)]] = "{"~repsep(arrow, ",")~"}" ^^ { case "{"~m~"}" ⇒ Map()++m}

    def arrow: Parser[(String, (String, String))] = word~":"~word~"→"~word ^^ {case f~":"~x~"→"~y ⇒ (f, (x, y))}

    def explain[T](pr: ParseResult[Result[T]]): Result[T] = {
      pr match {
        case Success(res, _) ⇒ res
        case e: NoSuccess ⇒ Result.error(s"Failed to parse: $e")
      }
    }

    def readGraph(input: CharSequence): Result[Graph] = {
      val pr: ParseResult[Result[Graph]] = parseAll(all, input)
      explain(pr)
    }

    def readGraph(input: Reader): Result[Graph] = explain(parseAll(all, input))
  }

  def read(input: Reader): Result[Graph] = (new Parser).readGraph(input)

  def read(input: CharSequence): Result[Graph] = (new Parser).readGraph(input)

  implicit class GraphString(val sc: StringContext) extends AnyVal {
    def graph(args: Any*): Graph = {
      val strings = sc.parts.iterator
      val expressions = args.iterator
      var buf = new StringBuffer(strings.next)
      while (strings.hasNext) {
        buf append expressions.next
        buf append strings.next
      }
      Graph.read(buf) match {
        case Good(g) ⇒ g
        case bad ⇒ 
          throw new InstantiationException(bad.errorDetails.mkString)
      }
    }
  }
}
